src/caching/backends.py

```diff
@@ -0,0 +1,141 @@
+# from: https://github.com/long2ice/fastapi-cache
+# Unfortunately, this library does not support caching of post requests (yet),
+# hence we use a slightly modified own version.
+import abc
+import time
+from asyncio import Lock
+from dataclasses import dataclass
+from typing import Dict, Optional, Tuple
+
+from databases import Database
+
+from lib.logger import create_logger
+
+logger = create_logger()
+
+
+class Backend:
+    @abc.abstractmethod
+    async def setup(self):
+        """Run intialization routines for the backend at application startup."""
+
+    @abc.abstractmethod
+    async def get_with_ttl(self, key: str) -> tuple[int, Optional[str]]:
+        """If not found, return None for the value. the returned ttl may be anything."""
+
+    @abc.abstractmethod
+    async def get(self, key: str) -> Optional[str]:
+        pass
+
+    @abc.abstractmethod
+    async def set(self, key: str, value: str, expire: int = None):
+        pass
+
+    @abc.abstractmethod
+    async def clear(self, key: str = None):
+        pass
+
+
+@dataclass
+class Value:
+    data: str
+    ttl_ts: int
+
+
+class InMemoryBackend(Backend):
+    _store: Dict[str, Value] = {}
+    _lock = Lock()
+
+    @property
+    def _now(self) -> int:
+        return int(time.time())
+
+    def _get(self, key: str):
+        v = self._store.get(key)
+        if v:
+            if v.ttl_ts < self._now:
+                del self._store[key]
+            else:
+                return v
+
+    async def setup(self):
+        pass  # nothing to do...
+
+    async def get_with_ttl(self, key: str) -> Tuple[int, Optional[str]]:
+        logger.info(f"get_with_ttl: {key}")
+        async with self._lock:
+            v = self._get(key)
+            if v:
+                return v.ttl_ts - self._now, v.data
+            return 0, None
+
+    async def get(self, key: str) -> Optional[str]:
+        logger.info(f"get {key}")
+        async with self._lock:
+            v = self._get(key)
+            if v:
+                return v.data
+
+    async def set(self, key: str, value: str, expire: int = None):
+        logger.info(f"set {key} {value} {expire}")
+        async with self._lock:
+            self._store[key] = Value(value, self._now + (expire or 0))
+
+    async def clear(self, namespace: str = None, key: str = None):
+        logger.info(f"clear {key}")
+        if namespace:
+            keys = list(self._store.keys())
+            for key in keys:
+                if key.startswith(namespace):
+                    del self._store[key]
+        elif key:
+            del self._store[key]
+
+
+class DatabaseBackend(Backend):
```

Comment:
Where is this being used?

Reply:
It wasn't up to now (will have to add this with the configuration via environment variables). It is a drop in replacement for the `InMemoryBackend` that is initialized at application startup. Actually, we could simply drop the `InMemoryBackend` as sqlite  kind of provides the same functionality with `sqlite://:memory:` as database url.

Reply:
Then yes, please, fuse both classes.

Reply:
Then yes, please, fuse both classes.

Reply:
removed the in memory part